File: solutions/prolog/raindrops/raindrops.pl
make_sound(N, "PlingPlangPlong") :- N mod 105 =:= 0, !.
make_sound(N, "PlangPlong") :- N mod 35 =:= 0, !.
make_sound(N, "PlingPlong") :- N mod 21 =:= 0, !.
make_sound(N, "Plong") :- N mod 7 =:= 0, !.
make_sound(N, "PlingPlang") :- N mod 15 =:= 0, !.
make_sound(N, "Plang") :- N mod 5 =:= 0, !.
make_sound(N, "Pling") :- N mod 3 =:= 0, !.
make_sound(N, Sound) :- number_string(N, Sound).

convert(N, Sounds) :- make_sound(N, Sounds).

File: solutions/nim/raindrops/raindrops.nim
const sounds = block:
  const factors = [
    (3, "Pling"),
    (5, "Plang"),
    (7, "Plong"),
  ]

  func calculateLength(a: openArray[(int, string)]): int =
    ## determine the product of integers in array of factors
    result = 1
    for item in a:
      result *= item[0]

  func generateLookup: array[calculateLength(factors), string] =
    ## setup raindrop sounds corresponding to factor combinations
    for n, s in result.mpairs():
      for (val, sound) in factors:
        if n mod val == 0:
          s.add sound

  generateLookup()

proc convert*(num: int): string =
  result = sounds[num mod sounds.len]
  if result == "":
    result = $num
